src/mbgl/style/expression/type_check.cpp

```diff
@@ -0,0 +1,121 @@
+#include <mbgl/style/expression/type_check.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+using namespace mbgl::style::expression::type;
+
+TypecheckResult typecheck(const Type& expected, const std::unique_ptr<Expression>& e) {
```

Comment:
👍  And I still think `Typename` should be eliminated as a type; a virtual `typecheck` method could be part of accomplishing that. The other part, I think, is replacing as many uses of `Typename` as possible with a set of overload signatures. The overload signature should include the return type, and expressions like `==`, `<`, `at` should have a set of overloads, one for each supported type. Not only will this reduce the need for `Typename`, it will force us to be more specific about which types these operators support and the definition for each supported type. For example, in the JS implementation I can see that `==` doesn't work correctly for `Color`, and `<` likely should not be defined for `Object`.

Then for the expression types that are still using `Typename`, define a custom `typecheck` method instead.